taichi/transforms/demote_operations.cpp

```diff
@@ -0,0 +1,85 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/program/program.h"
+
+TLANG_NAMESPACE_BEGIN
+
+// Demote Operations into pieces for backends to deal easier
+class DemoteOperations : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+  DelayedIRModifier modifier;
+
+  explicit AlgSimp(bool fast_math_)
+      : BasicStmtVisitor(), fast_math(fast_math_) {
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    auto lhs = stmt->lhs;
+    auto rhs = stmt->rhs;
+    if (stmt->op_type == BinaryOpType::floordiv) {
+      if (is_integral(rhs->element_type()) &&
+          is_integral(lhs->element_type())) {
+        // @ti.func
+        // def ifloordiv(a, b):
+        //     r = ti.raw_div(a, b)
+        //     if (a < 0) != (b < 0) and a and b * r != a:
+        //         r = r - 1
+        //     return r
```

Comment:
I would suggest not involving `b * r != a` since I don't think we can optimize it out easily.

What we need is:
```
a >= 0, b > 0: a / b
a >= 0, b < 0: (a - b + 1) / b
a <= 0, b > 0: (a - b + 1) / b
a <= 0, b < 0: a / b
```
So `((a < 0) == (b < 0) ? a : (a - b + 1)) / b` should be fine.

Reply:
It doesn't work for `a = 2, b = -3`:

```py
import taichi as ti


@ti.func
def ifloordiv(a, b):
    return (a if (a < 0) == (b < 0) else (a - b + 1)) / b


@ti.kernel
def func(a: int, b: int) -> int:
    return ifloordiv(a, b)


a, b = 2, -3
print(func(a, b), a // b)
```

This is non-trivial, ff2 change it iapr on your own.

Reply:
Oh, it should be
```
a >= 0, b > 0: a / b
a >= 0, b < 0: (a - b - 1) / b
a <= 0, b > 0: (a - b + 1) / b
a <= 0, b < 0: a / b
```
......
BTW when do we encounter `floordiv` at the frontend? Only when `ti.floordiv(a, b)`?